server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -352,24 +355,73 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
+                return currentMapper;
+            }
+            if (currentMapper.mapping().getRoot().subobjectsExplicit()) {
+                subobjects = Explicit.explicitBoolean(currentMapper.mapping().getRoot().isEnabled());
+            }
+        }
+
+        List<Mapping> parsedMappings = new LinkedList<>();
+        for (CompressedXContent mappingSource : mappingSources) {
+            Mapping mapping = parseMapping(type, mappingSource);
+            if (mapping.getRoot().subobjectsExplicit()) {
+                boolean explicitSubobjects = mapping.getRoot().subobjects();
+                if (subobjects != null && subobjects.value() != explicitSubobjects) {
+                    throw new MapperParsingException("Failed to parse mappings: contradicting subobjects settings provided");
+                }
+                subobjects = Explicit.explicitBoolean(explicitSubobjects);
+            }
+            parsedMappings.add(mapping);
+        }
+
+        if (subobjects != null && subobjects.value() != ObjectMapper.Defaults.SUBOBJECTS.value()) {
+            // we need to reparse as parsing is affected by the subobjects setting
+            parsedMappings.clear();
+            for (CompressedXContent mappingSource : mappingSources) {
+                parsedMappings.add(parseMapping(type, mappingSource, subobjects));
+            }
+        }
+
+        DocumentMapper ret = null;
+        for (Mapping parsedMapping : parsedMappings) {
+            ret = doMerge(reason, parsedMapping);
         }
-        synchronized (this) {
-            Mapping incomingMapping = parseMapping(type, mappingSource);
-            Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason);
-            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
-            // We should identify these cases reliably and save expensive serialization here
-            DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
-            if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-                return newMapper;
+        return ret;
+    }
+
+    public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
+        Explicit<Boolean> subobjects = null;
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null) {
+            if (currentMapper.mappingSource().equals(mappingSource)) {
+                return currentMapper;
+            }
+            if (currentMapper.mapping().getRoot().subobjectsExplicit()) {
+                subobjects = Explicit.explicitBoolean(currentMapper.mapping().getRoot().isEnabled());
             }
-            this.mapper = newMapper;
-            assert assertSerialization(newMapper);
+        }
+        Mapping incomingMapping = parseMapping(type, mappingSource, subobjects);
```

Comment:
Note to reviewer: this was formerly within the `synchronized` block but I assume that parsing doesn't really need to be synchronized and as far as I could tell it does seem to be safe to leave it out, but please take another look.

Reply:
Some `DataStreamIT` tests failure made me see that mapping parsing is not really a stateless operation as I initially thought. At least when it invokes [`metadataMappersSupplier.get()`](https://github.com/elastic/elasticsearch/blob/fdcecd9031521d9ffb08bea53097e232e48f5a80/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java#L104), it actually uses [`MapperService.getMetadataMappers()`](https://github.com/elastic/elasticsearch/blob/fdcecd9031521d9ffb08bea53097e232e48f5a80/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L210-L232), which is stateful as it depends on the `MapperService#mapper` field, which is being set every time [`MapperService.merge()`](https://github.com/elastic/elasticsearch/blob/fdcecd9031521d9ffb08bea53097e232e48f5a80/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L369) is invoked.
Therefore, not only I cannot extract mapping parsing outside of the `synchronized` block, it is logically incorrect to first parse all mappings and then merge them one-by-one.
So I'll close this PR and think of a different approach...